proposals/0000-class-method-headers.rst

```diff
@@ -0,0 +1,437 @@
+      f @b c @a :: (a, b, c) -> Bool
+
+   -- f :: forall b. forall c -> forall a. C a b c => (a, b, c) -> Bool
+   -- F :: forall b. forall c -> forall a.            (a, b, c) -> Bool
+
+These proposed binders would enable the programmer to control two aspects of
+type variable quantification:
+
+1. **Ordering.** The order of class variables in a method declaration would no
+   longer be tied to their order in the class declaration header.
+
+2. **Visibility**. By omitting the ``@``, the programmer could mark a type
+   variable visible, i.e. quantified with ``forall a ->``.
+
+Relevant extensions: ``TypeApplications``, ``RequiredTypeArguments``.
+
+Motivation
+------------
+
+At the moment, the signature of a class method is prefixed with class variables
+and a class constraint::
+
+  class C a b c where
+    f :: (a, b, c) -> Bool
+
+  -- f :: forall a b c. C a b c => (a, b, c) -> Bool
+  --      ^^^^^^^^^^^^^^^^^^^^^^^^
+  --      compiler-generated prefix
+
+
+The users have no way to guide generation of this prefix. In particular, the
+ordering of the class variables is inherited from the class declaration header.
+This matters at use sites with ``TypeApplications``::
+
+  f @Int :: forall b. C Int b c => (Int, b, c) -> Bool
+
+Here the type application ``@Int`` instantiates the outermost ``forall`` in the
+type of ``f``, namely ``forall a.`` (the first class variable).
+
+What if we, in the design of our API, wanted this type application to
+instantiate ``b``? ::
+
+  f @Int :: forall a. C a Int c => (a, Int, c) -> Bool
+
+Normally this would be done by changing the declaration of ``f`` in such a way
+that ``forall b.`` would become the outermost quantifier in the type of ``f``.
+However, Haskell lacks the syntax to do that in a class method declaration,
+where quantification over class variables is done entirely by the compiler.
+
+That is, we cannot specify that we'd prefer  ``forall b a c.`` instead of
+``forall a b c.`` in the type of ``f``. There are two workarounds:
+
+* Change the ordering of class variables in the class declaration header:
+  ::
+
+    class C b a c where   -- instead of:  class C a b c
+
+  This has the downside that now the class constructor has a different ordering
+  of variables, not just its method. Also, if there is more than one method,
+  all of them would be affected by such a change, even if our goal is to change
+  just one.
+
+* Create a wrapper function:
+  ::
+
+    fWrapper :: forall b a c. C a b c => (a, b, c) -> Bool
+    fWrapper @b @a @c = f @a @b @c
+
+  This means that users have to define instances using original method names,
+  but call the methods using the wrappers, which leads to more complicated
+  APIs.
+
+The problem is further exacerbated by the introduction of ``RequiredTypeArguments``
+(visible forall in types of terms). If we want to visibly quantify over some of
+class variables, there are also two workarounds:
+
+* Dummy type variable with an equality constraint:
```

Comment:
One interesting note about this workaround, which doesn't apply to the others, is that since arguments will still need to be in dependency order, this workaround will still have a place in cases where that's not desirable, as in e.g.
```haskell
ghci> (.:) :: forall a. a -> forall b -> a ~ b => a; x .: _ = x
ghci> 4 .: Float
4.0
```
(Not a class method, but illustrates the point I think.)